File: api/models/game.py
import random

from django.db import models

from api.events import on_event
from api.exceptions import NoMoreQuestions

GAME_STATES = (
    ('idle', 'idle'),
    ('started', 'started'),
    ('finished', 'finished'),
)


class Game(models.Model):
    """
    Game fields:
        - lang: string
        - state: string

    Game relations:
        - owner: Player
        - players: Player[]
        - question_master: Player
        - questions: Question[]
        - current_question: Question
        - choices: Choice[]
        - answers: AnsweredQuestion[]
        - turns: GameTurn[]

    Game methods:
        - get_history() -> GameTurn[]
        - add_player(player) -> None
        - remove_player(player) -> None
        - pick_question() -> None
        - deal_cards() -> None
        - get_propositions() -> AnsweredQuestion[] | None
    """

    lang = models.CharField(max_length=8)
    state = models.CharField(max_length=8, default='idle', choices=GAME_STATES)
    owner = models.ForeignKey('Player', related_name='owns', on_delete=models.CASCADE)
    question_master = models.ForeignKey('Player', blank=True, null=True, related_name='question_master_of', on_delete=models.CASCADE)
    current_question = models.ForeignKey('Question', blank=True, null=True, related_name='current', on_delete=models.CASCADE)

    def __str__(self):
        return ''.join(['Game #', str(self.id), ' (', self.state, ')'])

    def get_history(self):
        return self.turns.all()

    def add_player(self, player):
        self.players.add(player)
        on_event('game_joined', player)

    def remove_player(self, player):
        on_event('game_left', player)
        self.players.remove(player)

    def pick_question(self):
        available_questions = self.questions.filter(available=True)

        if available_questions.count() == 0:
            raise NoMoreQuestions

        self.current_question = random.choice(available_questions)
        self.current_question.available = False
        self.current_question.save()

    def deal_cards(self):
        for player in self.players.all():
            player.pick_card(self.choices)

    def get_propositions(self):
        if self.state != 'started':
            return None

        return self.answers.filter(question=self.current_question)


class GameTurn(models.Model):
    """
    GameTurn fields:
        - number: integer

    GameTurn relations:
        - game: Game
        - question_master: Player
        - winner: Player
        - question: Question
    """

    number = models.IntegerField()
    game = models.ForeignKey(Game, related_name='turns', on_delete=models.CASCADE)
    question_master = models.ForeignKey('Player', related_name='turns_as_question_master', on_delete=models.CASCADE)
    winner = models.ForeignKey('Player', related_name='turns_won', on_delete=models.CASCADE)
    question = models.OneToOneField('Question', related_name='turn', on_delete=models.CASCADE)

File: api/src/models/game.js
const Sequelize = require('sequelize');
const Op = Sequelize.Op;

module.exports = (sequelize, DataTypes) => {

  const Game = sequelize.define('Game', {
    lang: DataTypes.STRING,
    state: DataTypes.STRING,
    nbQuestions: DataTypes.INTEGER,
    cardsPerPlayer: DataTypes.INTEGER,
  }, {
    tableName: 'game',
  });

  Game.associate = function({ Question, Choice, Player, Answer, GameTurn }) {
    Game.belongsTo(Player, { as: 'owner', foreignKey: 'ownerId' });
    Game.belongsTo(Player, { as: 'questionMaster', foreignKey: 'questionMasterId' });
    Game.hasMany(Player, { as: 'players', foreignKey: 'gameId' });
    Game.hasMany(GameTurn, { as: 'turns', foreignKey: 'gameId' });
    Game.hasMany(Question, { as: 'questions', foreignKey: 'gameId' });
    Game.hasMany(Choice, { as: 'choices', foreignKey: 'gameId' });
    Game.belongsTo(Question, { as: 'currentQuestion', foreignKey: 'questionId' });
    Game.hasMany(Answer, { as: 'answers', foreignKey: 'gameId' });
    Game.belongsTo(Answer, { as: 'selectedAnswer', foreignKey: 'selectedAnswerId' });

    Game.defaultScope = {
      include: ['players', 'owner'],
    };
  };

  Game.prototype.getPlayState = async function() {
    if (this.state !== 'started')
      return null;

    if (this.selectedAnswerId)
      return 'end_of_turn';

    const propositions = await this.getPropositions();
    const playersCount = await this.countPlayers();

    if (propositions.length === playersCount - 1)
      return 'question_master_selection';

    return 'players_answer';
  };

  Game.prototype.createQuestions = async function(n) {
    const MasterQuestion = sequelize.model('masterquestion');
    const Question = sequelize.model('question');

    const mquestions = await MasterQuestion.findAll({
      where: { lang: this.lang },
      order: Sequelize.fn('RANDOM'),
      limit: n,
    });

    if (mquestions.length < n)
      throw new Error('not enough questions');

    return await Question.bulkCreate(mquestions.map(mq => {
      const values = mq.get();

      values.available = true;
      values.gameId = this.id;
      delete values.id;

      return values;
    }));
  };

  Game.prototype.createChoices = async function(n) {
    const MasterChoice = sequelize.model('masterchoice');
    const Choice = sequelize.model('choice');

    const mchoices = await MasterChoice.findAll({
      where: { lang: this.lang },
      order: Sequelize.fn('RANDOM'),
      limit: n,
    });

    if (mchoices.length < n)
      throw new Error('not enough choices');

    return await Choice.bulkCreate(mchoices.map(mc => {
      const values = mc.get();

      values.available = true;
      values.gameId = this.id;
      delete values.id;

      return values;
    }));
  };

  /**
   * Create the questions and choices from master data
   */
  Game.prototype.createCards = async function() {
    const questions = await this.createQuestions(this.nbQuestions);

    const p = await this.countPlayers();
    const b = questions.reduce((sum, q) => sum + q.getNbChoices(), 0);
    const n = this.cardsPerPlayer;
    const nbChoices = b * (p - 1) + n * p;

    await this.createChoices(nbChoices);
    // console.log('nbChoices: ' + nbChoices);
  };

  Game.prototype.getPropositions = async function(where = {}) {
    if (!this.questionId)
      return;

    return await this.getAnswers({
      where: {
        ...where,
        questionId: { [Op.eq]: this.questionId },
      },
      orderBy: ['place', 'id'],
      include: ['choices'],
    });
  };

  Game.prototype.toString = function() {
    return 'Game<#' + this.id + '>';
  };

  return Game;
};
